### DISCUSSION TITLE
Can we avoid `api` call on every route and call only on initial load ???!!!

### DISCUSSION THREAD

**Discussion by udayshankark-qp** [vercel/next.js#70313] [Help]
How to avoid api call on every route, can we just have the api call on initial load or any page and not on every route of the page

**Comment by icyJoseph** [comment]
Explain more, with pseudo-code perhaps?

**Reply by udayshankark-qp** [reply]
@icyJoseph 

So lets consider I have a page.tsx which is a server component


```
import ScreenTemplate from '@/components/screen/screenTemplate ';

export async function generateMetadata({ params, searchParams }: PageProps) {
	const { screen } = <Fetch api to get the screen response>;
	return <Generate the meta information using the screen response>
}

export default async function Page({
	params,
	searchParams,
}: PageProps): Promise<JSX.Element> {
	const { screen }: any =  <Fetch api to get the screen response>
	const seoSchemaDetails= <Generate Seo Schema Structured data>;

	return (
		<>
			<script
				type="application/ld+json"
				dangerouslySetInnerHTML={{ __html: JSON.stringify(seoSchemaDetails) }}
			/>
			<ScreenBody screen={screen} />
		</>
	);
}
```


And lets consider ScreenTemplate is the client component inside the page.tsx



```
import { useEffect, useState } from 'react';

export default function ScreenTemplate({
	screen
}: {
	screen: any;
}): JSX.Element {

const [clientScreen, setClientScreen] = useState();

useEffect(()=> {
// Inside this component also we do call the same screen api but using the authtoken (Since the api called in the server gives different response since it does have access to the auth token, so to get latest information we call the api in the client as well)
setClientScreen( <screen response from api>)
}, [])

const screenData = clientScreen ?? screen;

return <Display the template from `screenData` either from screen or clientScreen>
}
```





So here as you see in server component we call an Api which is same for fetching meta data and also for the screen template
In the client component also we call the same Api but now with auth token, since the response what we  get in server component would be either cached sometimes or the it would have difference response since its called without auth token.

so is there a way to either pass the Authtoken form the first route onwards to server component Api or somehow we need to avoid the Api on server component after initial load of the page, so that on every route it doesn't call server auth less Apis

**Reply by icyJoseph** [reply]
Right, I see, and the auth-token is stored in local storage or what? How come the server running context can't access it? A cookie for example, would be accessible server side.

**Reply by udayshankark-qp** [reply]
@icyJoseph 
Even I was thinking of accessing Cookies, but wont it also affect the server rendering
Since we will be reading cookies every time on render and it will might take some time load 

also on the initial load of the page the cookie might not be set as well, so there is possibility that the api might be called without auth token in server component

**Reply by icyJoseph** [reply]
I see, but then, doesn't it make sense for you to start with just fetching from the client? And have the server just ship the HTML shell of your app? Sort of CSR, mostly because of the cookie, not being set for some reason.

Otherwise, SSR, and doing it all server side should be fine, I know that on reload or whatever, re-running things can take time, but it will also be the case on the browser, since I guess, this data changes often, and it is bound to a user, so caching it might be a tedious task. 

Anyway, I think your original question was to like, once you are on the client, how to just call that API once. I'd say, look into SWR or React-Query, to pre-fill with server side data, and then fetch on the client, given certain conditions.

- https://swr.vercel.app/docs/advanced/understanding

Of course, you can do this manually with a `useState` hook to, and save the result, but I see none of your code, so I am just guessing what would fit best for you.
